lib/Storage/S3StorageClient.cpp

```diff
@@ -426,6 +439,23 @@ S3StorageClient::getWritableFile(llvm::StringRef Path,
 }
 
 llvm::Error S3StorageClient::commit() {
+  Pool.wait();
+
+  // NOTE: it is not safe to do this in a multi-threaded environment. We're ok
+  //       with it because right now the rest of the infrastructure does not
+  //       work multithreaded.
```

Comment:
Can you elaborate how this is not multi-thread safe?
AFAIU the `commit` method wasn't thread safe even before.
Accessing `PendingUploads` after `.wait()` should ensure no other threads are looking at the data in there.

Am I missing something?

Reply:
What I mean with that message is that, in an hypothetical situation where the revng api is accessed in a multi-threaded fashion, using `.wait` is not sufficent, an example sequence:
1. `.wait` finishes
2. Another thread starts saving, adding another element to `PendingUploads`
3. This thread starts iterating over `PendingUploads`, which includes (2) which should not be there

So in that case, the entire function should have a lock that prohibits `PendingUploads` from being manipulated while `commit` is running.

Reply:
> in an hypothetical situation where the revng api is accessed in a multi-threaded fashion

If this was a problem even before this changeset, it's not a problem.
Is this changeset making things worse in any way?

I think that, even when we'll be multithreaded, we'll be able to consider an "atomic transaction" the span of time from the first write until the invocation of `commit`.

Reply:
>  Is this changeset making things worse in any way?

No

> I think that, even when we'll be multithreaded, we'll be able to consider an "atomic transaction" the span of time from the first write until the invocation of commit.

Ok, got it, will drop the comment